parallel/runner.go

```diff
@@ -158,26 +143,112 @@ func (r *runner) Errors() map[int]error {
 func (r *runner) DoneWhenAllIdle(idleThresholdSeconds int) error {
 	for {
 		time.Sleep(time.Duration(idleThresholdSeconds) * time.Second)
-		for i := 0; i < r.maxParallel; i++ {
-			if !r.idle[i] {
-				break
+		allIdle := true
+		var e error
+		r.idle.Range(func(key, value interface{}) bool {
+			threadId, ok := key.(int)
+			if !ok {
+				e = errors.New("thread ID must be a number")
+				// this will break iteration
+				return false
+			}
+			threadIdle, ok := value.(bool)
+			if !ok {
+				e = errors.New("thread idle value must be a boolean")
+				// this will break iteration
+				return false
 			}
 
-			idleTimestamp, err := strconv.ParseInt(r.lastActive[i], 10, 64)
+			if !threadIdle {
+				allIdle = false
+				return false
+			}
+
+			lastActiveValue, _ := r.lastActive.Load(threadId)
+			threadLastActive, _ := lastActiveValue.(string)
+			idleTimestamp, err := strconv.ParseInt(threadLastActive, 10, 64)
 			if err != nil {
-				return errors.New("unexpected idle timestamp on consumer. err: " + err.Error())
+				e = errors.New("unexpected idle timestamp on consumer. err: " + err.Error())
+				return false
 			}
 
 			idleTime := time.Unix(idleTimestamp, 0)
 			if time.Now().Sub(idleTime).Seconds() < float64(idleThresholdSeconds) {
-				break
+				allIdle = false
+				return false
 			}
+			return true
+		})
+		if e != nil {
+			return e
+		}
 
-			// All consumers are idle for the required time.
-			if i == r.maxParallel-1 {
-				close(r.tasks)
-				return nil
-			}
+		// All consumers are idle for the required time.
+		if allIdle {
+			close(r.tasks)
+			return nil
 		}
 	}
 }
+
+func (r *runner) RunningThreads() int {
+	return r.runningThreads
+}
+
+func (r *runner) SetMaxParallel(newVal int) {
+	if newVal < 1 {
+		newVal = 1
+	}
+	if newVal == r.maxParallel {
+		return
+	}
+	if newVal > r.maxParallel {
```

Comment:
The following lop adds the new threads. Let's explain here the mechanism that causes threads to be removed, in case `newVal` is smaller than `r.maxParallel`.